packages/ember/tests/helpers/link_to_test/link_to_with_query_params_test.js

```diff
@@ -1,506 +1,577 @@
+      {{#link-to 'index' id='change-nothing'}}Index{{/link-to}}
+    `);
+    this.addTemplate('search', `
+      {{#link-to (query-params search='same') id='same-search'}}Index{{/link-to}}
+      {{#link-to (query-params search='change') id='change-search'}}Index{{/link-to}}
+      {{#link-to (query-params search='same' archive=true) id='same-search-add-archive'}}Index{{/link-to}}
+      {{#link-to (query-params archive=true) id='only-add-archive'}}Index{{/link-to}}
+      {{#link-to (query-params search='same' archive=true) id='both-same'}}Index{{/link-to}}
+      {{#link-to (query-params search='different' archive=true) id='change-one'}}Index{{/link-to}}
+      {{#link-to (query-params search='different' archive=false) id='remove-one'}}Index{{/link-to}}
+      {{outlet}}
+    `);
+    this.addTemplate('search.results', `
+      {{#link-to (query-params sort='title') id='same-sort-child-only'}}Index{{/link-to}}
+      {{#link-to (query-params search='same') id='same-search-parent-only'}}Index{{/link-to}}
+      {{#link-to (query-params search='change') id='change-search-parent-only'}}Index{{/link-to}}
+      {{#link-to (query-params search='same' sort='title') id='same-search-same-sort-child-and-parent'}}Index{{/link-to}}
+      {{#link-to (query-params search='same' sort='author') id='same-search-different-sort-child-and-parent'}}Index{{/link-to}}
+      {{#link-to (query-params search='change' sort='title') id='change-search-same-sort-child-and-parent'}}Index{{/link-to}}
+      {{#link-to (query-params foo='dog') id='dog-link'}}Index{{/link-to}}
+    `);
+
+    this.router.map(function() {
+      this.route('search', function() {
+        this.route('results');
+      })
     });
-  });
 
-  App.CarsController = Controller.extend({
-    queryParams: ['page'],
-    page: 1
-  });
+    this.add('controller:search', Controller.extend({
+      queryParams: ['search', 'archive'],
+      search: '',
+      archive: false
+    }));
+
+    this.add('controller:search.results', Controller.extend({
+      queryParams: ['sort', 'showDetails'],
+      sort: 'title',
+      showDetails: true
+    }));
+
+    this.visit('/').then(() => {
+      this.shouldNotBeActive(assert, '#cat-link');
+      this.shouldNotBeActive(assert, '#dog-link');
+    });
 
-  bootApplication();
+    this.visit('/?foo=cat').then(() => {
 
-  let carsController = container.lookup('controller:cars');
+      this.shouldBeActive(assert, '#cat-link');
+      this.shouldNotBeActive(assert, '#dog-link');
+    });
 
-  run(() => router.handleURL('/cars/create'));
+    this.visit('/?foo=dog').then(() => {
+      this.shouldBeActive(assert, '#dog-link');
+      this.shouldNotBeActive(assert, '#cat-link');
+      this.shouldBeActive(assert, '#change-nothing');
 
-  run(() => {
-    equal(router.currentRouteName, 'cars.create');
-    jQuery('#close-link').click();
-  });
+    });
+    this.visit('/search?search=same').then(() => {
```

Comment:
I mentioned some chaining of `visit` call stuff below. It applies here and maybe a few other spots as well. Basically because we currently have synchronous `visit` implementations, these test pass even though we haven't properly chained them. However in the future we may make this work async, and for that eventuality let's chain them. The fact that they are promises implies we should chain them to ensure sequential execution anyway.